### ISSUE TITLE
[Feature Request] Implement T9

### ISSUE THREAD

**Issue by maverick74** [AnySoftKeyboard/AnySoftKeyboard#1412]
T9 patented technology ( [https://worldwide.espacenet.com/publicationDetails/biblio?CC=US&NR=5818437&KC=&FT=E&locale=en_EP#](https://worldwide.espacenet.com/publicationDetails/biblio?CC=US&NR=5818437&KC=&FT=E&locale=en_EP#) ) is set to expire in next October 6th.

I'm not sure if one can work on the technology before the patent expires and release the implementation on the day the patent expires, or if one can only work on the technology after the patent expiration, but it would be great to have this ASAP

**Comment by TimMagee** [comment]
T9? As in the old system used on feature phones without QWERTY keyboards to speed up text entry from a numeric keypad? How exactly does that benefit ASK?

**Comment by maverick74** [comment]
It does not benefit ASK in anything!

But it does benefit users that, like me, have big fingers, have to have to write on the cell phone with two hands and that hates to have to use keyboards like TouchPal just because of this feature.

About the amount of users that does use this, I'm actually not the guy that can provide you with statistics but i can tell you that a lot of people around me uses this...

But maybe it's a "local thing".

(you managed to make me feel an old guy, however. From your talk i guess you never lived the good old Atari and ZX Spectrum days ahahaha... actually... i guess you never lived without a cellphone, have you? ahahah just messing with you :p)

**Comment by TimMagee** [comment]
>It does not benefit ASK in anything!
I have to disagree. Or maybe my previous question was the wrong one. I meant, "How does it benefit ASK _users_?"
If this benefits the users, it "benefits ASK" as the goal of ASK is to help the users input on their phones.

**Comment by maverick74** [comment]
Well, anyway you got your answer. :) At least from my POV...

**Comment by maverick74** [comment]
let me add that:

The compact keyboard layout is also great for me (it has big keys as well), but it also suffers from the same problem that, according to menny, is a consequence of T9 being patented ( see bug #834 ).

Still, let me clarify my suspect: @menny how difficult would you consider the implementation of T9 on ASK as it is right now? Would it be easy, hard, etc?

**Comment by maverick74** [comment]
Hurray!!! 

T9 not patented anymore :D

(sorry for the comment, but i believe it's good news. I've been waiting for this patent to expire for a long time)

**Comment by maverick74** [comment]
@menny please, could you consider include it in 1.9-r6 ?

**Comment by maverick74** [comment]
Any plans for this feature?

**Comment by llight3** [comment]
I would also love to see this happen, it's especially useful on smartphones with small screens or for searching contacts. Seems easy enough to implement since you already have a decent pradictive text feature you would only need to add the public algorithm for ambiguous input.

**Comment by maverick74** [comment]
@llight3 i also get the same feeling!
I believe this would be just a mere adjustment.
I think that recognizing that two keystrokes on the same key means two letters (and not one) would do the trick ( like mentioned in #834 )

**Comment by pawlosck** [comment]
Any progress? Any code added? I am waiting for this feature too.

**Comment by maverick74** [comment]
I've been waiting for this for years now (mentioned it many times in other bugs)...

I'm starting to feel this will not get implemented ever... :(

(very very disappointed about it...)

**Comment by pawlosck** [comment]
Probably, we have to wait for older developer (someone, who remember how to use phone without touch screeen) who implement this feature ;-).

**Comment by maverick74** [comment]
Damn!!! I'm not old!!! I'm not even in my 40s!!!
LOLOL

Now, jokes aside. I use T9 all the time! For example, I'm typing in T9 right now using TouchPal.

My reason to do so is because I have big fingers which don't do so well with a qwerty keyboard and because I also don't need to "travel so much space" to write a word (which might be ergonomically good).

It does work really well, if ask me. My problem with TouchPal is that it's not open-source and so I don't know what data they collect (beside getting all those boring ads)

**Comment by maverick74** [comment]
Well... 

Not that it may be important but since it may count for something: I'll Just to let "you" know I'm removing ASK as it's been a fairly frustrating wait for this feature.

Best wishes for the project.

(Feel free to close this issue)

**Comment by Merlignux** [comment]
Would also vote on this!

**Comment by harmtemolder** [comment]
I actually like TouchPal's approach, "T+". Kind of like T9, but based on QWERTY and two letters per key instead of three. I don't like the emoji circus that is TouchPal, though, so I'd rather see something like this in AnySoftKeyboard. I'm not an app developer, but as soon as someone starts working on this, I'd be happy to help think the implementation through

**Comment by maverick74** [comment]
@harmtemolder ASK also has that layout.
Problem is that since the "T9 predictability" / "Way of function" is not implemented it does not work like in TouchPal.

E.g.: You can't press 2x the same key for having two different letters. It always thinks your inserting the first letter from that key.

It's a real PITA!

**Comment by llight3** [comment]
This could be done if ASK implemented multitap gesture support (like how the shift button works) so it might be worth opening another feature request for that as that is something that would help users even outside of T9 for instance a multi tap could cycle through specific layers of a keyboard (EG English to numbers to Cyrillic ) or to show permantly the modifyers that you would normally get by holding said key. To my understanding this would also be better implemented at a lower level than T9. T9 could be a simple keyboard and would work like any other language however multi tap jestures for all keys would be a lower level feature that would then need to be exposed to the language  keyboard devs.

**Comment by maverick74** [comment]
@llight3  yes, you're probably right...
I only opened this one because when i mentioned i could not type "Hi" from the T9 keyboard layout the reason given to me was that T9 was patented (at the time) and that for that reason it could not be implemented!

Maybe what you suggest would solve the issue completely... i don't know...

**Comment by pawlosck** [comment]
I hope that someone will finally implement it.

**Comment by maverick74** [comment]
@pawlosck ... It's better that you find yourself a bed if you plan on waiting for it.

i like ASK a lot, but I ended up giving up on this... It's been a very frustrating wait... I would really like to replace TouchPal, but the fact is that I need a keyboard that works, and ASK ain't doing the thing - for me.

**Comment by pawlosck** [comment]
I am using original Samsung keyboard. Generally is ok, but privacy policy isn't ok for me. Some things are broken or I can't change it. Sometimes T9 is ON and sometimes is OFF and I don't know how to on permamently or On when is needed.

**Comment by zaskokus** [comment]
I'd like to vote for this feature as well. I'm stuck with TouchPal as other users, for same reasons. Implementation of T9 in there would be a spectacular improvement in usability.

**Comment by whoizit** [comment]
I use Smart Keyboard Trial for T9. I gave up TouchPal a long time ago because of ads and glitches.

**Comment by zaskokus** [comment]
> I use Smart Keyboard Trial for T9. I've long since quit the paid version of TouchPal cause agressive advertising and cpu load.

I know we went off-topic right now, but blocking internet traffic for TP keyboard app made it butter smooth and ads free.

**Comment by maverick74** [comment]
@zaskokus Yes, I've done the same thing with the same results. But I still don't trust it. And I would like very much to switch over to ASK. Even if I had to let go of some features.

Unfortunately, first T9 was under a patent, now it's no more. It became free in 2018. But still no one seems to care about this...

Even Emoji is more important than this... (I personally don't see anyone caring about emoji. But there you have it! You get emoji updates and nothing new T9. But maybe it's a "me and my friends thing" that don't use this emoji tendency)

Anyway, I still think it wouldn't be "a lot of work" to implement it... But then again, I might be wrong about it.

The only way out seems to be a do-it-yourself feature. Unfortunately, I'm not up to the task and I don't see anyone interested in it that can code...

So, TouchPal it is....... :(

**Comment by maverick74** [comment]
By the way, do anyone knows about any other FOSS keyboard that has T9? 

I've searched for it A LOT with on luck. F-droid, at least, has no alternative...

**Comment by zaskokus** [comment]
I remember I used to use another proper T9 keyboard like 2-3 years ago and I've a hard time finding it again.
But regradless of the outcome - I wish ASK has this implemented. The TP implementation + layout + feel + options is one of the best-ever attempts to create a fat-finger-proof super useful keyboard in the android world ever. I'd pay for a proper app if that ever existed.

**Comment by itzikiap** [comment]
I've found a keyboard, with similar layout to T9. It's  very buggy  when you type individual  keys , but  it works surprisingly well when gesture typing.
It's called "AnySoftKeyboard - Alternative Keyboards Pack".
But  yeah , it's  a bit frustrating to use…

**Comment by maverick74** [comment]
I think something really reliable would be needed.

For frustration there are a lot of other unavoidable things choose from

**Comment by TheSmallTeaBoi** [comment]
I really can't explain how useful t9 in a compact layout is, it makes typos basically non-existant.

**Comment by zaskokus** [comment]
@maverick74  @itzikiap @TheSmallTeaBoi 
So I've poked the internets a little, there's a cool project that's super tiny and provides all the T9 features we need: https://play.google.com/store/apps/details?id=kl.ime.oh "Multiling O Keyboard", it has some rough edges and convoluted settings, but once I went through it, it works much much better than jailed/sandboxed TouchPal.

**Comment by maverick74** [comment]
@zaskokus Yes, I've also heard of it recently.
I'll try it and see how it behaves.

Anyway, correct me if I'm wrong, it's sadly not open-source is it?

**Comment by zaskokus** [comment]
@maverick74 thats sadly true, but it's actually quite neat comparing to the pure awfullness that TouchPal is.

**Comment by TwiggyWan** [comment]
> @maverick74 @itzikiap @TheSmallTeaBoi So I've poked the internets a little, there's a cool project that's super tiny and provides all the T9 features we need: https://play.google.com/store/apps/details?id=kl.ime.oh "Multiling O Keyboard", it has some rough edges and convoluted settings, but once I went through it, it works much much better than jailed/sandboxed TouchPal.

Hi, how do you exactly set it up to have t9 prediction? I can find the '2ABC' layout but cannot make it predict (like older phones or touch pal)

**Comment by jonnyledger** [comment]
This app also has a proper t9 keyboard with predictive text, as well as other compact keyboards. It'd be great to have this functionality in an open source app
https://play.google.com/store/apps/details?id=app.minibytes.keyboard

**Comment by maverick74** [comment]
> This app also has a proper t9 keyboard with predictive text, as well as other compact keyboards. It'd be great to have this functionality in an open source app https://play.google.com/store/apps/details?id=app.minibytes.keyboard

It's just sad Tappy isn't OpenSource...

**Comment by Teloshav** [comment]
NOTE: Please do excuse me if the format/layout/everything of this comment thing is completely wrong, I have very little idea on how Github works, let alone different things like comments on issues.

TL;DR: I like this feature, and I would like to see if I could try add it myself. I am a complete beginner at this kind of thing so I hope that either I found out on my own through testing and/or a person(s) related to this project in depth could explain the theory/layout/feasibility behind potentially implementing such a feature.

----------------------------------------------

This is something I could get behind as well.

I've tried one of the T9-like keyboards for this app in the past, but it would have many issues when I would press it multiple times.

Just installed this app again after all that time and it seems to work way better now! I just don't know where to enable any potential predictive typing feature, if such a thing exists. (Using "EN Keypad = 16 keys English layout")

I realised that one of my old phones had this 'T9 dictionary' feature and I just forgot about it, as I never really mastered it, but I do recall coming across it, without not knowing what exactly it was, so I never tried mastering it as I didn't think much of it to read up more about it. (Yes, most of my typing was done the slow way... Would've been nice to learn that there was a fast place. :| Whoops!)

While I don't mind all too much about typing the 'slower' way with T9 (at least the normal predictions are helpful, even if they're not 'T9-styled'), I would like to see something like this come to life.

If I were to implement this feature myself, how would I go about doing so?

Do I make a module? Do I make a fork of the app and add the feature into the core app itself, like the dictionaries normally are? Any ideas?

I hope I can either learn the answer on my own and try see if I can make this feature come to life (don't get your hopes up, I've got just about no experience; I'll definitely give it my best shot if I could see myself actually pulling it off 😄) or have one of the developers/maintainers explain in a bit of detail on the app and this feature being a part of it — if their time permits — and then give it a shot if it seems like something I could have a crack at.

**Comment by maverick74** [comment]
@Teloshav  
a very small briefing :

= if you need to address directly to someone start by typing an @ and then typing the name of the intended person
= menny is the project owner, he has the last word
= About T9: the idea behind T9 was actually pretty simple:
-    Consider the word "dog":

       _the old way:_
       press 1 x [3] key, 3 x [6] key and 1 x [4] key

       _the T9 way:_
       press 1 x [3] key, **_1 x [6] key_** and 1 x [4] key

the predictive system should guess the word was "dog" (and give other options), but you would have pressed the 4 key just one time, instead of 3.

**Comment by Teloshav** [comment]
@maverick74 Thanks for letting me know about these things!

After reading about how people have used it, I seemed to have correctly guessed how it does work, great!

Now comes the question of developing it... 🤔

Should I address Menny about it? I don't want to be a nuisance, but I am not too sure where I should go or who I could talk to about the details of such a feature...

I have been using the AnySoftKeyboard T9 keyboard without this functionality and I can already see myself (and others!) enjoying the final piece of the puzzle to come together to fully implement T9 typing.

I'd be very happy to see if I could do it as I'll learn much in the process. Contributing to open source is something I've wanted to do for a long time, and this could be my first proper opportunity! 😁

- Teloshav

**Comment by joukewitteveen** [comment]
The part of T9 that needs to be implemented is the _generation_ of suggestions. Most other aspects of T9 are either already in-place, or irrelevant (T9 covers multiple patents, all expired). In fact, what needs to be implemented is a generalized T9, where a keyboard layout defines equivalence classes (i.e. the letters on each key of the selected keyboard layout) on the input 'alphabet'.

As far as I can tell, there are three main steps to the implementation. From easy to hard:
- [ ] Add a toggle somewhere in the UI to turn our variant of T9 on or off.
- [ ] In [SuggestImpl.java](https://github.com/AnySoftKeyboard/AnySoftKeyboard/blob/1.11-r1/ime/app/src/main/java/com/anysoftkeyboard/dictionaries/SuggestImpl.java#L251-L252), generalize prefix matching to matching modulo the equivalence class.
- [ ] Implement a [dictionary class](https://github.com/AnySoftKeyboard/AnySoftKeyboard/tree/1.11-r1/ime/dictionaries/src/main/java/com/anysoftkeyboard/dictionaries) backed by an in-memory search tree indexed by equivalence classes.

The difficulty lies in the fact that the equivalence classes are determined by the selected keyboard layout and hence the search tree needs to be built at runtime. For instance, there are currently at least two compact layouts: one for qwerty and one for dvorak. These layouts put up to two letters on each key, but do so in different combinations.

**Comment by Teloshav** [comment]
I see what you mean when you say that "most other aspects of T9 are either already in-place, or irrelevant" and that the main thing that needs to be implemented is the generation of suggestions.

Other than a few bugs I've come across, the T9 experience with AnySoftKeyboard, while tedius, is great compared to where it was when I first tried it out around a year or two ago.

Not sure if I'll be able to implement it myself, but I'll definitely give it a shot. :slightly_smiling_face:

Thanks for the little to-do list thingy!

- Teloshav

**Comment by maverick74** [comment]
I don't use ASK for years now but, if i remember correctly, my problem was with typing words that existed within the same keys e.g.: type: "Hi". We should be able to just press the "4" key 2 times and be done with it. 

The last time i tried it, i had to "play" with time configurations and wait to type the second letter, while on a "correct implementation of t9" you just type... no need to wait for it...

Would be good if you take a shoot at it @Teloshav

**Comment by Teloshav** [comment]
I don't know how to 'unreact' that rocket emoji I put...

Well that's not why I'm typing this. I'm going to start and try it out, wish me luck! :D

No matter what happens, this will be a good learning experience for me; so I am happy for this opportunity. 😀 

- Teloshav

**Comment by maverick74** [comment]
> I don't know how to 'unreact' that rocket emoji I put...

Just left-click on it!!!

> wish me luck! :D

Yeah, i REALLY do wish you the best of luck!!! I would really wish i could start using a decent T9 OpenSource Keyboard instead of TouchPal

(Make no mistake: **TouchPal works like a charm!!!** It's the perfect reference of how a _T9 keyboard_ should work!!! But it's closed source and i just don't trust it!)

**Comment by Teloshav** [comment]
I have literally no clue what I'm doing... BUT I DID THIS! HA! 😎🥳

https://user-images.githubusercontent.com/131397191/236890750-ce5902ab-94cf-4eba-8adf-4da8cda7376c.mp4

It's a slow start, spent some time looking around and finding out how these `.xml` file things work. I think I'm getting more familiar with this environment; only one way to tell though! 😅

I'm glad I have a bit of Flutter and Dart experience, so I'm not _completely_ lost, no doubt I'm still lost; this will be a learning process for me.

Yippee! (Thank you @joukewitteveen once again for this awesome To-Do list thing, I have no clue how this app works, let alone much Android apps in general! This has been very handy for me so far. :-] )

The JoukeWitteven 'T9 Suggestions'  To-Do List<sup><b>TM</b></sup>:
- [x] Add a toggle somewhere in the UI to turn our variant of T9 on or off.
- [ ] In [SuggestImpl.java](https://github.com/AnySoftKeyboard/AnySoftKeyboard/blob/1.11-r1/ime/app/src/main/java/com/anysoftkeyboard/dictionaries/SuggestImpl.java#L251-L252), generalize prefix matching to matching modulo the equivalence class.
- [ ] Implement a [dictionary class](https://github.com/AnySoftKeyboard/AnySoftKeyboard/tree/1.11-r1/ime/dictionaries/src/main/java/com/anysoftkeyboard/dictionaries) backed by an in-memory search tree indexed by equivalence classes.

The quest to add this T9 dictionary feature continues! (I just realised I should call it "T9 suggestions", I have done that now.)

<img width="344" alt="T9 Suggestions renaming" src="https://user-images.githubusercontent.com/131397191/236892903-84f34530-8cfc-4256-bb66-6a24b56c8908.png">

Hope this feature (if I manage to implement it fully, that is!) is going to be better than TouchPal! @maverick74 I'll start using that app to get a feel for what you're looking for and to get familiar with what I should try bring to AnySoftKeyboard. :)

 - Teloshav

**Comment by Teloshav** [comment]
### General thoughts/discussion:

I have been using Tappy as I couldn't get TouchPal working, and for the life of me I couldn't get [K9Keyboard](https://github.com/danielshorten/k9t9) working either... I stopped bothering as I have a GrapheneOS phone, so I thought my case would be too annoying to try cater towards (or might be impossible because of some security feature, which is something I don't mind happening personally, but is an issue I come across sometimes when using certain apps/features within apps).

> It's definitely the prediction dictionary.
I have disabled the multiple keystrokes wait (or something similar) and i do get the desired behavior, but it does not suggest the right words...

Hmmm... I'm glad you've pointed this out, and it's something I want to make sure I do right.

I'll have to spend more time thinking about the components of this feature, as the logic behind the feature (which seems simple at a quick glance) shows its insidious complexity! Any input about the nuances of this feature would be (and is) great.

---

### 'T9 suggestions' - Details for the settings of the AnySoftKeyboard app

When you enable T9 suggestions, some things should be changed (here's what I've noticed/noted so far):
- [ ] Taps on the same key are immediately registered as different letters.
> NOTE: I don't want this to happen with the top left button, the one with ",.?!'" on it. On Tappy, this behaviour is also enabled on that key and it makes no sense; I have to go to a different menu to put a full stop or anything other than a comma, it's frustrating. (I have no clue how to make this an indented bullet point. 🤷)
- [ ] The setting to make a delay for the repeated key press should be disabled/blanked out. (I'm not sure if the setting exists on AnySoftKeyboard, so this may be an irrelevant point. Will cross off if that's the case.)
- [ ] The gestures settings will be disabled/blanked out as well.
- [ ] The word suggestions/predictive dictionary gets replaced with this really cool T9 one that works like magic.
- [ ] Some other stuff that I can't recall right now (or might not even exist). (Why did I make this a checkbox? Lol)

---

### General details about the T9 predictive dictionary thingy

I need to try think about how this T9 predictive dictionary will work (or what I should call it x_x), it'll be the hardest thing to get working by FAR. I'm glad I have managed to break this feature into smaller chunks; it's starting to feel more approachable and I'm happy for that!! Divide and conquer! 

I'll just put the idea(s) I have for this dictionary thing here as it comes to mind (will be done via edits):
- I want it to be independent of any languages (that could be applied to this style of keyboard) and/or dictionaries of available words
- I want it to be functional with different types of compact layouts (say the second type of compact layout currently available that has 2-ish letters per key), and ideally generalised enough to work with other implementations of T9-style keyboards (meaning I'll have to analyse how AnySoftKeyboard works with different keyboard layouts, which could allow me to get 2 things done at once here. I was thinking about making my own keyboard layout that's more tailored for features/functionality I (and you all) may like to see!)

---

### Thinking to oneself aloud about the predictive dictionary algorithm thingy

The algorithm for predicting the word being typed is going to be a challenge, I don't know if there's any existing frameworks that could be used or not. Either way, I'm just trying to come up with my own logic, as I assume any predictive models that are really good are not openly available.

It needs to be able to narrow down dictionaries of immense size quickly, and also effectively, such that when you do type a word; not only will that word be valid, it would be the word you'd have actually intended on typing.

Take Tappy for example, when I type "I'm" (key 4 pressed once and key 6 pressed once), Tappy assumes I'm typing "in". It's annoying as it's a short word, so that could mean something else for trying to mitigate that.

Maybe I could make it such that it'll 'learn' what words you type more often, and account for that, meaning that if I type "I'm" more than "in", it'll account for that? Not sure, and that's WAY into the future. Just thinking ahead for now.

I feel a good thing to add would be a quick way to go to another suggested word as a more realistic 'fix' for that, especially given how hard it'd be for me to try build a learning algorithm thingy, if I could even do it at all!!

As someone wise long ago said, BABY STEPS OMG. I'm just going to try get that settings thing fully done, and also to find a way to get the "immediate" tapping thingy added as well.

---

### Closing notes

Sorry for such a HUGE message/essay (more an essay at this point lmao), this is my first time doing anything like this, and it only feels appropriate for me to talk about what I'm doing/how I'm doing it, as I and likely am doing a few things wrong. Also, it's pretty fun. 😁😁😁

Oh my goodness, I've had to add headings to my messages now... Give it a week and I'll need to add a table of contents for my messages. 🗿💀

The T9 train continues! 🚋 

- Teloshav

> PS: Also, is this an appropriate place for me to put such updates? I really enjoy this kind of thing, just letting anyone interested know on how it's going and all that; the thing is that I'm not sure if there may be another, more appropriate place to keep everyone updated, or if this issue is fine to do so. (My lack of Github 'etiquette' once again showing!)

**Comment by maverick74** [comment]
> I've tried tappy also (in fact, i think i've tried most (all?) T9 keyboards from the playstore and f-droid), but still think touchpal is better.
> 
> > Now that I see what you all want and now want it myself even more
> 
> Wanting something is a powerful force!!! Your video is an accomplishment!!! It's like it's already available ahaha
> 
> (by the way, there is also plans for a T9 in [FlorisBoard](https://github.com/florisboard/florisboard) - someday... and there is also [K9Keyboard](https://github.com/danielshorten/k9t9) - which i don't seem to be able to get to work... maybe looking at the code there can give you some clues)
> 
> Edit: I have reinstalled it... It's definitely the prediction dictionary. I have disabled the _multiple keystrokes wait_ (or something similar) and i do get the desired behavior, but it does not suggest the right words...

I don't know if this is the right place or not (i suppose so). But i personally do like this kind of updates! :)

About the amount of work... Wouldn't it be better to just start an independent T9 keyboard from scratch?! (ahahaha just joking, naturally)

**Comment by TwiggyWan** [comment]
what would be awesome is to put the t9 predictive engine in some kind of library for it to be included in the many different foss keyboard apps that currently exist..

**Comment by joukewitteveen** [comment]
> about the predictive dictionary algorithm thingy

At this stage, the 'predictive dictionary algorithm thingy' is really a lot less of a conceptual challenge than you think it is. The starting point should be a dictionary stored as a search tree indexed by layout-based equivalence classes. This is already mostly how the current dictionary classes work, except that you would introduce an alphabet hash function that is the identity function when T9 is off and maps letters to their keys when it is on. With this in place, you already get prevalence ranking from the existing code base. Further tweaks and optimizations should only be considered once the groundwork is in place.

**Comment by Teloshav** [comment]
WOW!! You all are quick to reply!

> what would be awesome is to put the t9 predictive engine in some kind of library for it to be included in the many different foss keyboard apps that currently exist..
@TwiggyWan that is something U 100% would love to do. (Still need to make it first, but if I do manage to do it!)

> Wouldn't it be better to just start an independent T9 keyboard from scratch?!
@maverick74 💀 🥲 That's a sure fire way to drive a wild Teloshav clinically insane.

> At this stage, the 'predictive dictionary algorithm thingy' is really a lot less of a conceptual challenge than you think it is.
@joukewitteveen Thank you once again for the more technical details about how I could do this! Reading it all at has some of it flying over my head, but I'm going to spend a lot of time on this. I'll get more familiar with the terms and stuff you're using as time goes on. The way you're talking about it though is giving me a lot of confidence, so thank you!

> Further tweaks and optimizations should only be considered once the groundwork is in place.
@joukewitteveen Of course. :-)

**Comment by jonnyledger** [comment]
Regarding https://github.com/danielshorten/k9t9 - judging from the authors reddit posts, it looks like it's for phones with physical keyboards (if anyone like me was wondering why nothing is displayed on-screen after installing it).

If you were going to lift a T9 prediction engine from anywhere, it'd probably be from the "dumb phone" community - there seems to be a handfull of T9 projects for android phones with physical keyboards floating around.

**Comment by Teloshav** [comment]
@maverick74 Yes it worked!

I spent so much time looking through the settings and failing to find where the T9 keyboard settings was.

So for anyone who's looking for the TouchPal T9 phone pad keyboard thing settings and somehow got here by accident, here's where the setting is:

- Open the app.
- Get the keyboard installed so that you can type with the full size keyboard.
- On the top left hand side of the keyboard, you can see two symbols. Out of those two symbols, pick the one on the right. (looks like a globe)
- Then select the option `PHONEPAD`!

First impressions:
THIS IS GREAT.

HOW CAN IT FIX MY TYPOS, I NEED TO FIND OUT.

I can see why you like this app so much @maverick74! The thing I like about it the most is that the amount of times I need to re-type a word because I accidentally pressed the wrong key (or accidentally pressed the same key twice when I wasn't intending on doing so) has gone down drastically.

Keep in mind that this is in a few minutes of using it! I can imagine the impact of that on the long term when being used daily.

My testing and analysis of keyboards will be on TouchPal now, and no longer will be with Tappy. I'm glad I can get this app working with no permissions enabled at least!

I'm intrigued!

- Teloshav

**Comment by maverick74** [comment]
> HOW CAN IT FIX MY TYPOS, I NEED TO FIND OUT.

I'm not sure what you mean...

If you saved the word by mistake, the next time it shows up as a suggestion (on the suggestions bar), just long press on it and you will be able to edit it (you can also go to  _Settings > Prediction > Manage user dictionary_   [**NOTE: the path translations may not be 100% correct**]).

if it's another case, please clarify/exemplify what happens :)



> I can imagine the impact of that on the long term when being used daily.

Yeah... this "thing", after a few months, presents the right words i intended even if i press one or two wrong keys (possibly based on the context and on the words i use the most). It does that for both Portuguese and English at the same time.

It's a really good keyboard (just like old T9 on Nokias)!

The only problem is that it's closed source and that it keeps sending data to various countries (i had to block it...)

**Comment by Teloshav** [comment]
@maverick74 

>if it's another case, please clarify/exemplify what happens :)

Oh! Sorry I meant about how GOOD it is!

If I mistype ONCE on Tappy, it'll completely flop.

TouchPal, on the other hand, is incredible!

It still knows what words I want to use if I mistype once or even twice! It's completing words long before I finish typing them; it almost feels like it's reading my mind! (Now that I've had more time since I originally typed this message, this is very true. The predictive engine on TouchPal is spectacular!)

> It's a really good keyboard (just like old T9 on Nokias)!

Never used the Nokia keyboards, the only one I had used to have the full keypad. But from what you're saying, it sounds like I'm not missing out on too much! (I have used other T9 keyboards, but but it seems that the Nokia T9 keyboards are THE benchmark for a good T9 keyboard.)

Once again, seeing how much better this is than even Tappy is makes me want this kind of prediction engine open source! I'm not sure how good I can make it, but I'm _going_ to make it.

I did take a bit of a break from doing this as I do have much work to attend to outside of this, but I will be doing more work on analysing both the settings and the dictionaries of the app.

Looking forward to making some more progress on this feature again! :D

- Teloshav

(I had this typed out for some time, but somehow _just_ realised I never sent the message... How? Your guess is as good as mine!)

**Comment by maverick74** [comment]
@Teloshav How's it going?

**Comment by Teloshav** [comment]
@maverick74 I'm ashamed to say that I haven't made much progress. My studies for the next month or so are clear so the amount of obligations for the time being has gone down exponentially.

This means I can once again get back to working on this feature, I apologise for the slow progress! 😓

- Teloshav

**Comment by maverick74** [comment]
**No pressure!!!** The issue has 5 years already.
No problem if it takes another 5 ahahaha

**Comment by Teloshav** [comment]
@maverick74 Thanks for the patience. :)

**Comment by jonnyledger** [comment]
I don't know if anyone has spotted this but tt9 (which is primarily for hardware keypads) actually has an onscreen keyboard as a beta option that can be enabled in the settings. Prediction is pretty decent as well. 

https://github.com/sspanak/tt9

**Comment by maverick74** [comment]
> I don't know if anyone has spotted this but tt9 (which is primarily for hardware keypads) actually has an onscreen keyboard as a beta option that can be enabled in the settings. Prediction is pretty decent as well.

Was not aware of it... will try to test :) 
Thank you :)

**Comment by pawlosck** [comment]
I tested. Generally ok, but buttons are too small for me and buttons should be moved to right.

@maverick74 Why you can't install? I downloaded apk from github if I remember well.

**Comment by maverick74** [comment]
> @maverick74 Why you can't install? I downloaded apk from github if I remember well.

I didn't get it from github, but from f-droid! Still, it should be the same as the IzzyOnDroid repo, IIRC, drinks directly from github...

Anyway, i could not install it. I'll try again when i have more time... 

Does it have a GUI settings somewhere? (maybe i'm confusing with K9Keyboard... at least one of these do no have GUI Settings...)

**Comment by pawlosck** [comment]
Android show me this app as "traditional T9". I will send you later screenshots from this app and what to choose to enable this keyboard.

**Comment by pawlosck** [comment]
You are welcome. I had to back to Samsung keyboard, because buttons are too small for me. I will try to report improvement to change it and then I will use it.

**Comment by maverick74** [comment]
> You are welcome. I had to back to Samsung keyboard, because buttons are too small for me. I will try to report improvement to change it and then I will use it.

I'm having really big hopes that i can finally ditch touchpal for this.
It's a very basic keyboard but it's doing it's job pretty well so far and being open allows me to trust it!!!
(i'm going to test it during next days)
